### PR TITLE
Add possessive quantifiers to avoid catastrophic backtracking

### PR DESCRIPTION
**Pull request by l0rinc**
Fixes the crash in https://github.com/openai/tiktoken/issues/245 by prohibiting the regex engine from backtracking catastrophically via [possessive quantifiers](https://www.regular-expressions.info/possessive.html).

image

Interestingly these possesives make the encoding a lot faster again in `fancy-regex`.

Before this change (but with large byte pair merge PR cherry-picked):
```
num_threads: 1, num_bytes: 98379553
tiktoken 	11,946,036 bytes / s
tiktoken 	11,961,343 bytes / s
tiktoken 	11,995,846 bytes / s
tiktoken 	11,951,263 bytes / s
tiktoken 	11,983,405 bytes / s
```
Same, with these changes applied:
```
num_threads: 1, num_bytes: 98379553
tiktoken 	14,511,827 bytes / s
tiktoken 	14,638,134 bytes / s
tiktoken 	14,644,029 bytes / s
tiktoken 	14,729,030 bytes / s
tiktoken 	14,666,903 bytes / s
```
Updating the regex libs makes it a tiny bit faster still:
```
num_threads: 1, num_bytes: 98379553
tiktoken 	14,485,590 bytes / s
tiktoken 	14,854,049 bytes / s
tiktoken 	14,891,086 bytes / s
tiktoken 	14,843,007 bytes / s
tiktoken 	14,874,520 bytes / s
```

This is almost 2x faster than [before any of the optimizations](https://github.com/openai/tiktoken/pull/234).

-------

Opened an issue for increasing the [default backtrack limit](https://github.com/fancy-regex/fancy-regex/blob/bf2c807447f72ee20ae839e0f8cb3a06fc79982c/src/lib.rs#L407), see: https://github.com/fancy-regex/fancy-regex/issues/134, but it shouldn't be necessary here anymore.

### CODE REVIEW

File: tests/test_encoding.py [hunk 1]
```diff
@@ -11,6 +11,20 @@
 from .test_helpers import ENCODING_FACTORIES, MAX_EXAMPLES
 
 
+@pytest.mark.skip(reason="Takes a really long time to finish, but was added to reproduce a crash.")
+@pytest.mark.parametrize("make_enc", ENCODING_FACTORIES)
+def test_extremely_big_encoding(make_enc: Callable[[], tiktoken.Encoding]):
+    enc = make_enc()
+    for c in ["^", "0", "a", " ", "\n", "'s"]:
```
Comment (paplorinc) [inline tests/test_encoding.py:18]: all of these are stressing different parts of the regex, makin sure none have catastrophic backtracking

File: tests/test_encoding.py [hunk 2]
```diff
@@ -11,6 +11,20 @@
...
+        print(f"Validating `{c}`")
+
+        big_value = "^" * 1000000
+        assert big_value == enc.decode(enc.encode(big_value))
+
+        big_value = " " + big_value
```
Comment (paplorinc) [inline tests/test_encoding.py:24]: space if often optional beginning, this way the backtracking can reach the space - let's test that as well

File: tiktoken_ext/openai_public.py [hunk 3]
```diff
@@ -20,7 +20,7 @@ def gpt2():
         # The pattern in the original GPT-2 release is:
         # r"""'s|'t|'re|'ve|'m|'ll|'d| ?[\p{L}]+| ?[\p{N}]+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+"""
         # This is equivalent, but executes faster:
-        "pat_str": r"""'(?:[sdmt]|ll|ve|re)| ?\p{L}+| ?\p{N}+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+""",
+        "pat_str": r"""'(?:[sdmt]|ll|ve|re)| ?\p{L}++| ?\p{N}++| ?[^\s\p{L}\p{N}]++|\s+(?!\S)|\s++""",
```
Comment (paplorinc) [inline tiktoken_ext/openai_public.py:23]: except for `\\s+(?!\\S)` all groups should be possessive, no need to step back on match failure

File: tiktoken_ext/openai_public.py [hunk 4]
```diff
@@ -82,7 +82,7 @@ def cl100k_base():
     }
     return {
         "name": "cl100k_base",
-        "pat_str": r"""'(?i:[sdmt]|ll|ve|re)|[^\r\n\p{L}\p{N}]?+\p{L}+|\p{N}{1,3}| ?[^\s\p{L}\p{N}]++[\r\n]*|\s*[\r\n]|\s+(?!\S)|\s+""",
+        "pat_str": r"""'(?i:[sdmt]|ll|ve|re)|[^\r\n\p{L}\p{N}]?+\p{L}++|\p{N}{1,3}+| ?[^\s\p{L}\p{N}]++[\r\n]*+|\s*[\r\n]|\s+(?!\S)|\s++""",
```
Comment (paplorinc) [inline tiktoken_ext/openai_public.py:85]: seems the cl100k also had some backtracking problems, these possessives improve the situation considerably (e.g. in Java these aren't necessary, see https://github.com/knuddelsgmbh/jtokkit/pull/87)

File: tests/test_encoding.py [hunk 5]
```diff
@@ -11,6 +11,20 @@
 from .test_helpers import ENCODING_FACTORIES, MAX_EXAMPLES
 
 
+@pytest.mark.skip(reason="Takes a really long time to finish, but was added to reproduce a crash.")
```
Comment (paplorinc) [inline tests/test_encoding.py:14]: should be reenabled after https://github.com/openai/tiktoken/pull/239 is merged

File: tests/test_encoding.py [hunk 6]
```diff
@@ -11,6 +11,20 @@
...
+@pytest.mark.parametrize("make_enc", ENCODING_FACTORIES)
+def test_extremely_big_encoding(make_enc: Callable[[], tiktoken.Encoding]):
+    enc = make_enc()
+    for c in ["^", "0", "a", " ", "\n", "'s"]:
+        print(f"Validating `{c}`")
+
+        big_value = "^" * 1000000
+        assert big_value == enc.decode(enc.encode(big_value))
```
Comment (paplorinc) [inline tests/test_encoding.py:22]: roundtrip

File: tests/test_encoding.py [hunk 7]
```diff
@@ -11,6 +11,23 @@
 from .test_helpers import ENCODING_FACTORIES, MAX_EXAMPLES
 
 
+@pytest.mark.skip(reason="Takes a really long time to finish, but was added to reproduce a crash.")
+@pytest.mark.parametrize("make_enc", ENCODING_FACTORIES)
+def test_extremely_big_encoding(make_enc: Callable[[], tiktoken.Encoding]):
+    enc = make_enc()
+    for c in ["^", "0", "a", "'s"]: # TODO " ", "\n" are still failing
```
Comment (paplorinc) [inline tests/test_encoding.py:18]: I have tried atomic grouping here (and a few other tricks), but couldn't make it work.
So all-whitespace humonguous inputs will still fail, everything else is likely fixed

File: tests/test_encoding.py [hunk 8]
```diff
@@ -11,6 +11,23 @@
...
+        print(f"Validating `{c}`")
+
+        big_value = c * 1_000_000
+        assert big_value == enc.decode(enc.encode(big_value))
+
+        big_value = " " + big_value
+        assert big_value == enc.decode(enc.encode(big_value))
+
+        big_value = big_value + "\n"
```
Comment (paplorinc) [inline tests/test_encoding.py:27]: some groups require a newline at the end, stress those paths as well

File: tiktoken_ext/openai_public.py [hunk 9]
```diff
@@ -6,6 +6,11 @@
 FIM_SUFFIX = "<|fim_suffix|>"
 ENDOFPROMPT = "<|endofprompt|>"
 
+# The pattern in the original GPT-2 release is:
+# r"""'s|'t|'re|'ve|'m|'ll|'d| ?[\p{L}]+| ?[\p{N}]+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+"""
+# This is equivalent, but executes faster:
+_legacy_splitter_regex = r"""'(?:[sdmt]|ll|ve|re)| ?\p{L}++| ?\p{N}++| ?[^\s\p{L}\p{N}]++|\s+(?!\S)|\s++"""
```
Comment (paplorinc) [inline tiktoken_ext/openai_public.py:12]: The whitespaces can't be possessive (it needs to step back when encountering a non-shitespace), I have tried atomic grouping here, tried possessives with positive lookeahed, but couldn't make it work.
So all-whitespace humonguous inputs will still fail, everything else is likely fixed

File: src/lib.rs [hunk 10]
```diff
@@ -417,7 +418,7 @@ impl CoreBPE {
         special_tokens_encoder: HashMap<String, Rank>,
         pattern: &str,
     ) -> PyResult<Self> {
-        let regex = Regex::new(pattern)
+        let regex = RegexBuilder::new(pattern).backtrack_limit(100_000).build()
```
Comment (paplorinc) [inline src/lib.rs:421]: doesn't work for values bigger than a million - see https://github.com/fancy-regex/fancy-regex/issues/134

I've set it lower for now, hoping we'll be able to fix the whitespace problem

File: tiktoken_ext/openai_public.py [hunk 11]
```diff
@@ -82,7 +84,7 @@ def cl100k_base():
     }
     return {
         "name": "cl100k_base",
-        "pat_str": r"""'(?i:[sdmt]|ll|ve|re)|[^\r\n\p{L}\p{N}]?+\p{L}+|\p{N}{1,3}| ?[^\s\p{L}\p{N}]++[\r\n]*|\s*[\r\n]|\s+(?!\S)|\s+""",
+        "pat_str": r"""'(?i:[sdmt]|ll|ve|re)|[^\r\n\p{L}\p{N}]?+\p{L}++|\p{N}{1,3}+| ?[^\s\p{L}\p{N}]++[\r\n]*+|\s++$|\s*[\r\n]|\s+(?!\S)|\s""",
```
Comment (paplorinc) [inline tiktoken_ext/openai_public.py:87]: seems the cl100k also had some backtracking problems, these possessives improve the situation considerably (e.g. in Java these aren't necessary, see https://github.com/knuddelsgmbh/jtokkit/pull/87)

File: tests/test_encoding.py [hunk 12]
```diff
@@ -11,6 +11,22 @@
 from .test_helpers import ENCODING_FACTORIES, MAX_EXAMPLES
 
 
+@pytest.mark.parametrize("make_enc", ENCODING_FACTORIES)
+def test_extremely_big_encoding(make_enc: Callable[[], tiktoken.Encoding]):
+    enc = make_enc()
+    for c in ["^", "0", "a", "'s", " ", "\n"]:
```
Comment (paplorinc) [inline tests/test_encoding.py:17]: stressing different parts of the regex, makin sure none have catastrophic backtracking

File: tiktoken_ext/openai_public.py [hunk 13]
```diff
@@ -6,6 +6,11 @@
 FIM_SUFFIX = "<|fim_suffix|>"
 ENDOFPROMPT = "<|endofprompt|>"
 
+# The pattern in the original GPT-2 release is:
+# r"""'s|'t|'re|'ve|'m|'ll|'d| ?[\p{L}]+| ?[\p{N}]+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+"""
+# This is equivalent, but executes faster:
+_legacy_splitter_regex = r"""'(?:[sdmt]|ll|ve|re)| ?\p{L}++| ?\p{N}++| ?[^\s\p{L}\p{N}]++|\s++$|\s+(?!\S)|\s"""
```
Comment (paplorinc) [inline tiktoken_ext/openai_public.py:12]: The whitespaces can't be possessive (it needs to step back when encountering a non-whitespace), but we can rule out the offending bactracking case by adding a possessive trailing whitespace check.

File: Cargo.toml [hunk 14]
```diff
@@ -12,7 +12,7 @@ crate-type = ["cdylib"]
 pyo3 = { version = "0.20.0", features = ["extension-module"] }
 
 # tiktoken dependencies
-fancy-regex = "0.11.0"
-regex = "1.8.3"
+fancy-regex = "0.13.0"
+regex = "1.10.3"
```
Comment (paplorinc) [inline Cargo.toml:16]: not absolutely necessary, but adds a tiny speed increase

File: src/lib.rs [hunk 15]
```diff
@@ -417,7 +418,7 @@ impl CoreBPE {
         special_tokens_encoder: HashMap<String, Rank>,
         pattern: &str,
     ) -> PyResult<Self> {
-        let regex = Regex::new(pattern)
+        let regex = RegexBuilder::new(pattern).backtrack_limit(10_000).build()
```
Comment (paplorinc) [inline src/lib.rs:421]: after this change we should never backtract catastrophically - and if we do, this will warn us early

**Review by hauntsaninja — APPROVED** [review]
Thanks!

**Comment by tmm1** [comment]
Was the backtrack limit reverted intentionally in 05e66e8? https://github.com/openai/tiktoken/commit/05e66e8db7ef220d3c0b1aafbee5af289345684b#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759L421-R438

Was there a regression?

**Comment by hauntsaninja** [comment]
Yes, it was reverted intentionally. There are OpenAI internal encodings where setting the limit caused issues.

**Comment by l0rinc** [comment]
Thanks for checking @tmm1, @hauntsaninja.

File: tiktoken_ext/openai_public.py (hunk 11 above)
Comment (tmm1) [reply tiktoken_ext/openai_public.py:87]: can we collapse the `\s+(?!\S)` too? it should be equivalent to `\s+$` no?

Comment (l0rinc) [reply tiktoken_ext/openai_public.py:87]: It sounds reasonable, but check the tests to understand why they're not the same

**Pull request merged**